raw-ipa-lib/src/report.rs

```diff
@@ -0,0 +1,137 @@
+use crate::threshold::DecryptionKey as ThresholdDecryptionKey;
+use crate::threshold::{Ciphertext, RistrettoPoint};
+
+use std::collections::HashMap;
+use std::fmt;
+
+pub struct EncryptedMatchkeys {
+    match_keys: HashMap<String, Ciphertext>,
+}
+
+impl PartialEq for EncryptedMatchkeys {
+    fn eq(&self, other: &Self) -> bool {
+        any_matches(self.match_keys.values(), &other.match_keys.values())
+    }
+}
+
+impl EncryptedMatchkeys {
+    pub fn count_matches(&self, other: &Self) -> usize {
+        n_matches(self.match_keys.values(), &other.match_keys.values())
+    }
+
+    pub fn from_matchkeys(match_keys: HashMap<String, Ciphertext>) -> EncryptedMatchkeys {
+        EncryptedMatchkeys { match_keys }
+    }
+
+    pub fn threshold_decrypt(
+        &self,
+        matchkey_decrypt: &ThresholdDecryptionKey,
+    ) -> EncryptedMatchkeys {
+        let partially_decrypted_matchkeys = self
+            .match_keys
+            .iter()
+            .map(|(p, emk)| (p.to_string(), matchkey_decrypt.threshold_decrypt(*emk)))
+            .collect();
+        EncryptedMatchkeys::from_matchkeys(partially_decrypted_matchkeys)
+    }
+
+    pub fn decrypt(&self, matchkey_decrypt: &ThresholdDecryptionKey) -> DecryptedMatchkeys {
+        let decrypted_matchkeys = self
+            .match_keys
+            .iter()
+            .map(|(p, emk)| (p.to_string(), matchkey_decrypt.decrypt(*emk)))
+            .collect();
+        DecryptedMatchkeys::from_matchkeys(decrypted_matchkeys)
+    }
+}
+
+impl fmt::Debug for EncryptedMatchkeys {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self.match_keys)
+    }
+}
+
+pub struct DecryptedMatchkeys {
+    match_keys: HashMap<String, RistrettoPoint>,
+}
+
+impl PartialEq for DecryptedMatchkeys {
+    fn eq(&self, other: &Self) -> bool {
+        any_matches(self.match_keys.values(), &other.match_keys.values())
+    }
+}
+
+impl DecryptedMatchkeys {
+    pub fn count_matches(&self, other: &Self) -> usize {
```

Comment:
This looks like a function that you might only want to run for testing purposes.  You can move the implementation to the test module (the test module is able to access private fields), or you could do this:

```suggestion
    #[cfg(test)]
    pub fn count_matches(&self, other: &Self) -> usize {
```

I might prefer this in the test code, as then you don't have to deal with dead_code warnings and the like.